From ac607351aef320bc31cf3df65f82d51f2ac90916 Mon Sep 17 00:00:00 2001
From: Mystro256 <alexjnewt@hotmail.com>
Date: Tue, 18 Jan 2022 12:50:05 -0500
Subject: [PATCH] Use FHS compliant install

FHS does not allow installing to "amdgcn", but "LIBDIR/amdgcn" appears
to be acceptable to the FHS and distros. This logic uses GNUInstallDirs
to detect the LIBDIR instead of hardcoding it to "lib".

E.g. Fedora/Redhat use lib64 instead of lib.
---
 cmake/OCL.cmake                        | 3 ++-
 cmake/Packages.cmake                   | 2 +-
 test/compile/CMakeLists.txt            | 2 ++
 test/compile/RunConstantFoldTest.cmake | 1 +
 4 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/cmake/OCL.cmake b/cmake/OCL.cmake
index 773c6f6..3252082 100644
--- a/cmake/OCL.cmake
+++ b/cmake/OCL.cmake
@@ -10,6 +10,7 @@
 # configure_file and we are included from multiple places. To get around this
 # we `file(WRITE)` a file with an @variable reference and `configure_file` it.
 cmake_policy(SET CMP0053 OLD)
+include(GNUInstallDirs)
 
 if (WIN32)
   set(EXE_SUFFIX ".exe")
@@ -46,7 +47,7 @@ set (BC_EXT .bc)
 set (LIB_SUFFIX ".lib${BC_EXT}")
 set (STRIP_SUFFIX ".strip${BC_EXT}")
 set (FINAL_SUFFIX "${BC_EXT}")
-set (INSTALL_ROOT_SUFFIX "amdgcn/bitcode")
+set (INSTALL_ROOT_SUFFIX "${CMAKE_INSTALL_LIBDIR}/amdgcn/bitcode")
 
 # Set `inc_options` to contain Clang command-line for include directories for
 # current source directory.
diff --git a/cmake/Packages.cmake b/cmake/Packages.cmake
index 07c60eb..f128365 100644
--- a/cmake/Packages.cmake
+++ b/cmake/Packages.cmake
@@ -16,7 +16,7 @@ configure_file(AMDDeviceLibsConfig.cmake.in
   @ONLY)
 
 
-set(install_path_suffix "amdgcn/bitcode")
+set(install_path_suffix "${CMAKE_INSTALL_LIBDIR}/amdgcn/bitcode")
 
 # Generate the install-tree package.
 # We do not know the absolute path to the intall tree until we are installed,
diff --git a/test/compile/CMakeLists.txt b/test/compile/CMakeLists.txt
index 9af0b1a..3fe348b 100644
--- a/test/compile/CMakeLists.txt
+++ b/test/compile/CMakeLists.txt
@@ -4,6 +4,7 @@
 ## This file is distributed under the University of Illinois Open Source
 ## License. See LICENSE.TXT for details.
 ##===--------------------------------------------------------------------------
+include(GNUInstallDirs)
 
 if(TARGET FileCheck)
   set(FILECHECK_BIN $<TARGET_FILE:FileCheck>)
@@ -30,6 +31,7 @@ function(add_compile_test test_name func_name script test_cpu extra_check_prefix
      -DFILECHECK_BIN=${FILECHECK_BIN}
      -DOUTPUT_FILE=output.${name}.${test_cpu}.s
      -DINPUT_FILE=${CMAKE_CURRENT_SOURCE_DIR}/${func_name}.cl
+     -DAMDGCN_BITCODES=${PROJECT_BINARY_DIR}/${CMAKE_INSTALL_LIBDIR}/amdgcn/bitcode
      -DTEST_CPU=${test_cpu}
      -DEXTRA_CHECK_PREFIX=${extra_check_prefixes}
      -P ${script})
diff --git a/test/compile/RunConstantFoldTest.cmake b/test/compile/RunConstantFoldTest.cmake
index 5424690..1d4241a 100644
--- a/test/compile/RunConstantFoldTest.cmake
+++ b/test/compile/RunConstantFoldTest.cmake
@@ -16,6 +16,7 @@ execute_process(COMMAND
   -target amdgcn-amd-amdhsa -mcpu=${TEST_CPU}
   -Xclang -finclude-default-header
   --rocm-path=${BINARY_DIR}
+  --rocm-device-lib-path=${AMDGCN_BITCODES}
   -mllvm -amdgpu-simplify-libcall=0
   -o ${OUTPUT_FILE} ${INPUT_FILE}
   RESULT_VARIABLE CLANG_RESULT
-- 
2.39.2

